### ISSUE TITLE
Souffle interpreter cannot ingest -2^31

### ISSUE THREAD

**Issue by varming** [souffle-lang/souffle#276]
Souffle doesn't handle a simple program using -2^31.

Given
```
.decl foo(x0: String, x1: number) output printsize
foo("bar", -2147483648).
```
Souffle fails with:
```
Error: integer constant must be in range [0, 2147483647] in file minimal.db at line 2
foo("bar", -2147483648).
------------^------------
1 errors generated, evaluation aborted
```
This is surprising as the user manual (https://github.com/souffle-lang/souffle/wiki/user-manual) states that the allowed integer input range is: [ -2^31 .. 2^31 -1 ].

**Comment by psubotic** [comment]
@b-scholz -- is anyone looking into this? Or has it been resolved?

**Comment by b-scholz** [comment]
I had a quick look into this issue today. The problem is that -2^31 is really not the constant -2^31. It is semantically -(2^31) and hence it is out of range because only positive numbers between 0 and 2^31-1 are permitted. 

We could extend the scanner to permit negative constants. However, this leads to some anomalies such as ```x-1``` won't parse anymore because it would be interpreted as identifier x followed by number -1, i.e., the minus symbol is the sign of the constant -1 rather than the minus operation. We would force the programmer to add a blank between constant 1 and the operation minus, i.e., ```x - 1``` could be parsed. 

There is a nice write-up about this issue
http://stackoverflow.com/questions/34894434/flex-matching-signed-int-vs-addition-subtraction

I am not sure whether we should give the scanner context as explained in post or whether we should leave it as it is.

**Comment by varming** [comment]
A simpler solution is to let the NUMBER token contain a string and convert the string to the native integer type in the parser where you know the context.

**Comment by b-scholz** [comment]
That is not so easy because numbers can be either binary, hexadecimal, quads etc. Hence, the lexicographical analysis for numbers would need to be redone in the parser.

**Comment by lyndonhenry** [comment]
This issue still currently remains due to the explanation given in its parent issue #285, and so must be reopened again.

**Comment by lyndonhenry** [comment]
Closed as per pull request #294.
